proposals/record-structs.md

```diff
@@ -0,0 +1,303 @@
+Members are synthesized unless a member with a "matching" signature is declared in the record struct body or
+an accessible concrete non-virtual member with a "matching" signature is inherited.
+Two members are considered matching if they have the same
+signature or would be considered "hiding" in an inheritance scenario.
+
+It is an error for a member of a record struct to be named "Clone".
+
+It is an error for an instance field of a record struct to have an unsafe type.
+
+A record struct is not permitted to declare a destructor.
+
+The synthesized members are as follows:
+
+### Equality members
+
+The synthesized equality members are similar as in a record class (`Equals` for this type, `Equals` for `object` type, `==` and `!=` operators for this type),\
+except for the lack of `EqualityContract`, null checks or inheritance.
+
+The record struct implements `System.IEquatable<R>` and includes a synthesized strongly-typed overload of `Equals(R other)` where `R` is the record struct.
+The method is `public`.
+The method can be declared explicitly. It is an error if the explicit declaration does not match the expected signature or accessibility.
+
+If `Equals(R other)` is user-defined (not synthesized) but `GetHashCode` is not, a warning is produced.
+
+```C#
+public bool Equals(R other);
+```
+
+The synthesized `Equals(R)` returns `true` if and only if for each instance field `fieldN` in the record struct
+the value of `System.Collections.Generic.EqualityComparer<TN>.Default.Equals(fieldN, other.fieldN)` where `TN` is the field type is `true`.
+
+The record struct includes synthesized `==` and `!=` operators equivalent to operators declared as follows:
+```C#
+public static bool operator==(R r1, R r2)
+    => r1.Equals(r2);
+public static bool operator!=(R r1, R r2)
+    => !(r1 == r2);
+```
+The `Equals` method called by the `==` operator is the `Equals(R other)` method specified above. The `!=` operator delegates to the `==` operator. It is an error if the operators are declared explicitly.
+
+The record struct includes a synthesized override equivalent to a method declared as follows:
+```C#
+public override bool Equals(object? obj);
+```
+It is an error if the override is declared explicitly. 
+The synthesized override returns `other is R && Equals((R)other)` where `R` is the record struct.
+
+The record struct includes a synthesized override equivalent to a method declared as follows:
+```C#
+public override int GetHashCode();
+```
+The method can be declared explicitly.
+
+A warning is reported if one of `Equals(R)` and `GetHashCode()` is explicitly declared but the other method is not explicit.
+
+The synthesized override of `GetHashCode()` returns an `int` result of a deterministic function combining the values of `System.Collections.Generic.EqualityComparer<TN>.Default.GetHashCode(fieldN)` for each instance field `fieldN` with `TN` being the type of `fieldN`.
```

Comment:
Can/should we doc if deterministic has guarantees across multiple invocations of the compiler, or new versions of the compiler? Or if it is just deterministic within a single invocation of the compiler?

Reply:
We didn't for records (which this spec is largely based on), but I don't remember if we discussed that question then. 
In general, I think we try to make as few promises as possible with regards to hashing. Does `Combine` make such promise (stable hash over multiple version)?

---
In reply to: [553563047](https://github.com/dotnet/csharplang/pull/4307#discussion_r553563047) [](ancestors = 553563047)

Reply:
>  I think we try to make as few promises as possible 

Agreed.  bt i would go further.  I would explicitly state that we do *not* promise this.  As opposed to being silent on this.